File: android/app/src/main/java/com/tacmap/localization/LocalizedMessage.kt
package com.tacmap.localization

/** In-memory display state. Diagnostic details remain verbatim; never persist this as a wire format. */
data class LocalizedMessage internal constructor(
    val id: String?,
    val fallback: String,
    val arguments: List<String>,
    private val nestedArguments: Map<Int, LocalizedMessage> = emptyMap(),
) {
    val text: String get() = id?.let { L10n.message(it, fallback, *arguments.mapIndexed { index, value -> nestedArguments[index]?.text ?: value }.toTypedArray()) } ?: fallback

    fun withArgument(index: Int, message: LocalizedMessage): LocalizedMessage {
        require(index in arguments.indices)
        return copy(nestedArguments = nestedArguments + (index to message))
    }

    companion object {
        fun literal(text: String): LocalizedMessage = LocalizedMessage(null, text, emptyList())
    }
}

File: ios/TacticalMaps/Util/LocalizedMessage.swift
import Foundation

/// In-memory display state. Resolve app copy when shown; keep diagnostic details verbatim.
struct LocalizedMessage: Equatable {
    let id: String?
    let fallback: String
    let arguments: [String]

    static func literal(_ text: String) -> LocalizedMessage {
        LocalizedMessage(id: nil, fallback: text, arguments: [])
    }

    var text: String {
        guard let id else { return fallback }
        return L10n.resolveMessage(id, fallback: fallback, arguments: arguments)
    }
}
